caduceus_distill/distill.py

```diff
@@ -31,58 +37,87 @@
 def _filter_non_specific_nucleotides_and_batch(
     student_logits: torch.Tensor,
     teacher_logits: torch.Tensor,
+    student_emb: torch.Tensor | None,
+    teacher_emb: torch.Tensor | None,
     input_ids: torch.Tensor,
-) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+) -> tuple[
+    torch.Tensor, torch.Tensor, torch.Tensor | None, torch.Tensor | None, torch.Tensor
+]:
+    """
+    This function filters out the examples where the label is the non-specific nucleotide (PAD token).
+    """
     mask = input_ids != CADUCEUS_PAD_TOKEN_ID
     # 2d mask is going to collapses some dimensions
     student_logits = student_logits[mask]
     teacher_logits = teacher_logits[mask]
     input_ids = input_ids[mask]
+    student_emb = student_emb[mask] if student_emb is not None else None
+    teacher_emb = teacher_emb[mask] if teacher_emb is not None else None
 
     assert student_logits.ndim == 2
     assert teacher_logits.ndim == 2
+    assert student_emb.ndim == 2 if student_emb is not None else True
+    assert teacher_emb.ndim == 2 if teacher_emb is not None else True
     assert input_ids.ndim == 1
 
-    return student_logits, teacher_logits, input_ids
+    return student_logits, teacher_logits, student_emb, teacher_emb, input_ids
 
 
 def distillation_loss(
+    *,
     student_logits: torch.Tensor,
     teacher_logits: torch.Tensor,
+    student_emb: torch.Tensor | None,
+    teacher_emb: torch.Tensor | None,
     input_ids: torch.Tensor,
-    *,
-    temperature: float = 4.0,
-    alpha: float = 0.8,
-) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    temperature: float,
+    alpha_soft: float,
+    alpha_sim: float,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
     """
-    Calculate combined distillation loss (KL divergence with temperature scaling + cross-entropy with hard targets).
+    Calculate combined distillation loss.
 
-    Args:
+    Parameters:
         student_logits: Raw logits from student model [B, T, V]
         teacher_logits: Raw logits from teacher model [B, T, V]
+        student_emb: Hidden states from student model [B, T, D]
+        teacher_emb: Hidden states from teacher model [B, T, D]
         input_ids: Ground truth token ids (hard targets) [B, T]
         temperature: Temperature for softening probability distributions
-        alpha: Weight for distillation loss (1-alpha for hard loss)
-
-    Returns:
-        Weighted sum of distillation and hard target loss
+        alpha_soft: Weight for soft targets loss (KL divergence)
+        alpha_sim: Weight for hidden state similarity loss (cosine embedding loss)
     """
     assert temperature > 0, "Temperature must be positive"
-    assert 0 <= alpha <= 1, "Alpha must be in [0, 1]"
+    assert 0 <= alpha_soft <= 1, "alpha_soft must be in [0, 1]"
+    alpha_hard = 1.0 - alpha_soft
```

Comment:
Shouldn't `alpha_hard + alpha_soft + alpha_sim == 1.0`? 🤔

Reply:
> Can you articulate why we would want to enforce sum to 1.0?

Mostly a convention to have weighted sums have float factors that are [0, 1] and sum to one. That way, the weights can be interpreted as relative contributions to the loss, rather than (arbitrary?) absolute weights.

Not a big deal though.